File: flutter/lib/models/message.dart
/// Who wrote a message in a thread.
enum MessageAuthor { me, them, system }

/// What kind of content a message carries.
///
/// CareConnect's governing rule is that anything communicated through sound
/// must also be communicated visually or in text. The non-text kinds below
/// therefore all carry a readable payload:
///
/// * [voicemail] renders its transcript, never a bare "play" button.
/// * [videoMessage] states whether a caption track is attached.
/// * [alert] renders a visible banner with an icon and a word.
enum MessageKind { text, voicemail, videoMessage, alert }

/// Delivery state of an outgoing message.
///
/// The state is always rendered as a word next to its icon so it is never
/// carried by a tick colour alone.
enum DeliveryStatus {
  sending('Sending'),
  sent('Sent'),
  delivered('Delivered'),
  read('Read');

  const DeliveryStatus(this.label);

  final String label;
}

/// A single entry in a conversation with a contact.
class Message {
  const Message({
    required this.id,
    required this.contactId,
    required this.author,
    required this.body,
    required this.sentAt,
    this.kind = MessageKind.text,
    this.status = DeliveryStatus.sent,
    this.mediaLabel,
    this.hasCaptions = false,
  });

  final String id;
  final String contactId;
  final MessageAuthor author;

  /// The readable payload. For a [MessageKind.voicemail] this is the
  /// transcript; for a [MessageKind.videoMessage] it is the caption summary.
  final String body;

  final DateTime sentAt;
  final MessageKind kind;
  final DeliveryStatus status;

  /// Short description of the attached media, e.g. "Voicemail · 0:34".
  final String? mediaLabel;

  /// Whether an attached video carries a synchronized caption track
  /// (WCAG 2.2 success criterion 1.2.2).
  final bool hasCaptions;

  bool get isMine => author == MessageAuthor.me;
  bool get isSystem => author == MessageAuthor.system;

  /// True when the message body stands in for audio the user cannot hear.
  bool get isTextAlternative =>
      kind == MessageKind.voicemail || kind == MessageKind.videoMessage;

  /// Heading shown above the readable payload for media messages.
  ///
  /// Returns `null` for a plain text message, which needs no heading.
  String? get alternativeHeading {
    switch (kind) {
      case MessageKind.voicemail:
        return 'Transcript';
      case MessageKind.videoMessage:
        return hasCaptions ? 'Captions available' : 'Captions unavailable';
      case MessageKind.text:
      case MessageKind.alert:
        return null;
    }
  }

  /// The sentence assistive technology announces for this message.
  String semanticLabel(String contactName) {
    final String who = switch (author) {
      MessageAuthor.me => 'You',
      MessageAuthor.them => contactName,
      MessageAuthor.system => 'CareConnect alert',
    };
    final StringBuffer buffer = StringBuffer(who);
    final String? heading = alternativeHeading;
    if (heading != null) buffer.write(', $heading');
    buffer.write(': $body');
    if (isMine) buffer.write('. ${status.label}');
    return buffer.toString();
  }

  Message copyWith({
    String? body,
    DeliveryStatus? status,
    MessageKind? kind,
    String? mediaLabel,
    bool? hasCaptions,
    DateTime? sentAt,
  }) {
    return Message(
      id: id,
      contactId: contactId,
      author: author,
      body: body ?? this.body,
      sentAt: sentAt ?? this.sentAt,
      kind: kind ?? this.kind,
      status: status ?? this.status,
      mediaLabel: mediaLabel ?? this.mediaLabel,
      hasCaptions: hasCaptions ?? this.hasCaptions,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is Message && other.id == id);

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'Message($id, ${author.name}, ${kind.name})';
}

File: mobile/src/models/message.ts
/** Who wrote a message in a conversation. */
export type MessageAuthor = 'me' | 'them' | 'system';

/**
 * What kind of content a message carries.
 *
 * CareConnect's governing rule is that anything communicated through sound must
 * also be communicated visually or in text. The non-text kinds below therefore
 * all carry a readable payload: a voicemail renders its transcript, a video
 * states whether a caption track is attached, and an alert renders a visible
 * banner with an icon and a word.
 */
export type MessageKind = 'text' | 'voicemail' | 'videoMessage' | 'alert';

/**
 * Delivery state of an outgoing message. Always rendered as a word next to its
 * tick, so it is never carried by a tick colour alone.
 */
export type DeliveryStatus = 'sending' | 'sent' | 'delivered' | 'read';

export const deliveryStatusLabels: Record<DeliveryStatus, string> = {
  sending: 'Sending',
  sent: 'Sent',
  delivered: 'Delivered',
  read: 'Read',
};

/** A single entry in a conversation with a contact. */
export interface Message {
  id: string;
  contactId: string;
  author: MessageAuthor;
  /**
   * The readable payload. For a voicemail this is the transcript; for a video
   * message it is the caption summary.
   */
  body: string;
  /** Milliseconds since the epoch, so a message is trivially serialisable. */
  sentAt: number;
  kind?: MessageKind;
  status?: DeliveryStatus;
  /** Short description of the attached media, e.g. "Voicemail · 0:34". */
  mediaLabel?: string;
  /**
   * Whether an attached video carries a synchronized caption track
   * (WCAG 2.2 success criterion 1.2.2).
   */
  hasCaptions?: boolean;
}

export const isMine = (message: Message): boolean => message.author === 'me';
export const isSystem = (message: Message): boolean => message.author === 'system';
export const kindOf = (message: Message): MessageKind => message.kind ?? 'text';

/** True when the message body stands in for audio the user cannot hear. */
export function isTextAlternative(message: Message): boolean {
  const kind = kindOf(message);
  return kind === 'voicemail' || kind === 'videoMessage';
}

/**
 * Heading shown above the readable payload for media messages.
 *
 * Returns `null` for a plain text message, which needs no heading.
 */
export function alternativeHeading(message: Message): string | null {
  switch (kindOf(message)) {
    case 'voicemail':
      return 'Transcript';
    case 'videoMessage':
      return message.hasCaptions ? 'Captions available' : 'Captions unavailable';
    default:
      return null;
  }
}

/** The sentence assistive technology announces for this message. */
export function messageSemanticLabel(message: Message, contactName: string): string {
  const who =
    message.author === 'me'
      ? 'You'
      : message.author === 'system'
        ? 'CareConnect alert'
        : contactName;

  let label = who;
  const heading = alternativeHeading(message);
  if (heading) label += `, ${heading}`;
  label += `: ${message.body}`;
  if (isMine(message)) {
    label += `. ${deliveryStatusLabels[message.status ?? 'sent']}`;
  }
  return label;
}

/** The most recent message in a conversation, or null for an empty one. */
export function latestOf(thread: Message[]): Message | null {
  return thread.length === 0 ? null : thread[thread.length - 1];
}

/**
 * A one-line preview of the conversation for the contact list.
 *
 * Media messages preview their readable payload — the transcript or caption
 * summary — because a preview that just said "Voicemail" would leave a deaf
 * user with no idea whether it mattered.
 */
export function previewOf(thread: Message[]): string {
  const latest = latestOf(thread);
  if (!latest) return 'No messages yet';
  const prefix = isMine(latest) ? 'You: ' : '';
  const heading = alternativeHeading(latest);
  const body = heading ? `${heading} — ${latest.body}` : latest.body;
  return `${prefix}${body}`;
}

/**
 * Messages waiting for a reply.
 *
 * Counts the incoming messages at the end of a conversation — the ones that
 * arrived after the user last wrote back. A conversation the user has already
 * answered shows no badge even though it is full of incoming messages, which is
 * what the prototype's badges show.
 *
 * CareConnect's own alerts are skipped: they are not somebody waiting.
 */
export function waitingCount(thread: Message[]): number {
  let count = 0;
  for (let i = thread.length - 1; i >= 0; i -= 1) {
    const message = thread[i];
    if (isMine(message)) break;
    if (isSystem(message)) continue;
    count += 1;
  }
  return count;
}

/** Sorts a conversation oldest-first without mutating the input. */
export function chronological(thread: Message[]): Message[] {
  return [...thread].sort((a, b) => a.sentAt - b.sentAt);
}

/** True when `message` is the first of its calendar day within `thread`. */
export function startsNewDay(thread: Message[], index: number): boolean {
  if (index === 0) return true;
  const previous = new Date(thread[index - 1].sentAt);
  const current = new Date(thread[index].sentAt);
  return (
    previous.getFullYear() !== current.getFullYear() ||
    previous.getMonth() !== current.getMonth() ||
    previous.getDate() !== current.getDate()
  );
}
